extensions/devservices/deployment/src/main/java/io/quarkus/devservices/deployment/DevServicesProcessor.java

```diff
@@ -0,0 +1,228 @@
+        }
+        context.reset(
+                new ConsoleCommand('c', "Show dev services containers", null, () -> {
+                    List<DevServiceDescription> descriptions = buildServiceDescriptions(devServicesResults,
+                            devServicesLauncherConfig);
+                    StringBuilder builder = new StringBuilder();
+                    builder.append("\n\n")
+                            .append(RED + "==" + RESET + " " + UNDERLINE + "Dev Services" + NO_UNDERLINE)
+                            .append("\n\n");
+                    for (DevServiceDescription devService : descriptions) {
+                        printDevService(builder, devService, true);
+                        builder.append("\n");
+                    }
+                    System.out.println(builder);
+                }),
+                new ConsoleCommand('g', "Follow dev services logs to the console",
+                        new ConsoleCommand.HelpState(() -> logForwardEnabled ? GREEN : RED,
+                                () -> logForwardEnabled ? "enabled" : "disabled"),
+                        this::toggleLogForwarders));
+    }
+
+    private List<DevServiceDescription> buildServiceDescriptions(List<DevServicesResultBuildItem> devServicesResults,
+            Optional<DevServicesLauncherConfigResultBuildItem> devServicesLauncherConfig) {
+        // Fetch container infos
+        Set<String> containerIds = devServicesResults.stream()
+                .map(DevServicesResultBuildItem::getContainerId)
+                .filter(Objects::nonNull)
+                .collect(Collectors.toSet());
+        Map<String, Container> containerInfos = fetchContainerInfos(containerIds);
+        // Build descriptions
+        Set<String> configKeysFromDevServices = new HashSet<>();
+        List<DevServiceDescription> descriptions = new ArrayList<>();
+        for (DevServicesResultBuildItem buildItem : devServicesResults) {
+            configKeysFromDevServices.addAll(buildItem.getConfig().keySet());
+            descriptions.add(toDevServiceDescription(buildItem, containerInfos.get(buildItem.getContainerId())));
+        }
+        // Sort descriptions by name
+        descriptions.sort(Comparator.comparing(DevServiceDescription::getName));
+        // Add description from other dev service configs as last
+        if (devServicesLauncherConfig.isPresent()) {
+            Map<String, String> config = new HashMap<>(devServicesLauncherConfig.get().getConfig());
+            for (String key : configKeysFromDevServices) {
+                config.remove(key);
+            }
+            if (!config.isEmpty()) {
+                descriptions.add(new DevServiceDescription("Other Dev Services", null, config));
+            }
+        }
+        return descriptions;
+    }
+
+    private Map<String, Container> fetchContainerInfos(Set<String> containerIds) {
+        if (!isDockerWorking.getAsBoolean() || containerIds.isEmpty()) {
```

Comment:
If you flip this if condition (i.e. check containerIds.isEmpty() first) it will fix the memory leak. It is also a better design, as we should not be initializing the docker client if the project does not actually have any docker based dev services.

This is an existing leak that only affects QuarkusUnitTest (so is not something that really affects end users). Basically the testcontainers docker client is never closed, and it hangs onto the class loader forever. This change means that every test in the devmode module now opens a docker client, and this is never cleaned up so it runs out of memory.

We should look at some way of fixing this (possibly making testcontainers parent first), but it is an existing issue so does not have to happen with this PR (if you added a dep that used dev services to this module it would fail the same way).

Reply:
Thanks for the explanation @stuartwdouglas. This totally makes sense now.